### PR TITLE
[PM-11350] Use shared expiration year normalization util function

### PR DESCRIPTION
**Pull request by jprusik**
## 🎟️ Tracking

PM-11350

## 📔 Objective

We don’t currently enforce a normalized format for stored, user-entered values of expiry year (e.g. credit cards). Because of this, values like “00” are ambiguous, and invalid values like “0302", "awdawdawd", “40000”, and “4.5” are accepted.

Screenshot 2024-08-26 at 1 02 41 PM

These changes aim to normalize user submitted values to a YYYY format (given the context of expiration) so that downstream logic doesn’t have ensure proper handling these cases and to more authoritatively establish the full date to be stored.

Note, format normalization during new/edit cipher views happens upon save/submit (not on field value change).

## 📸 Screenshots

**browser client**

https://github.com/user-attachments/assets/36545025-0e48-4910-ae08-a7c6f198555d

**browser client with component refresh**

https://github.com/user-attachments/assets/d20dbb4d-4d9d-4aac-9a5c-44bfd406b6f5

**web client**

https://github.com/user-attachments/assets/e9162e5d-e277-4390-ba81-ac1ee6eaaefe

**desktop client**

https://github.com/user-attachments/assets/9dd73564-4e87-4b6f-89f5-ac1e4887d31f

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: libs/common/src/vault/utils.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  // Exit early if year is already formatted correctly
+  if (/^\d{4}$/.test(expirationYear)) {
+    return expirationYear as Year;
+  }
+
+  // For safety, because even input[type="number"] will allow decimals
+  expirationYear = expirationYear.replace(/[^\d]/g, "");
+
+  // given the context of payment card expiry, a year character length of 3, or over 4
+  // is more likely to be a mistake than an intentional value for the far past or far future.
```
Comment (jprusik) [inline libs/common/src/vault/utils.ts:25]: Because we have and maintain the context of usage intent (expiry year), we do not need to support _all_ valid year formats (e.g. "334", "203054", "-3653", "0"), since they aren't relevant

File: libs/common/src/vault/utils.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  if (yearInput !== null && expirationYear.length !== 4) {
+    const paddedYear = ("00" + expirationYear).slice(-2);
+    const currentCentury = `${new Date().getFullYear()}`.slice(0, 2);
```
Comment (jprusik) [inline libs/common/src/vault/utils.ts:29]: We could get more clever with this re: calculation around the turn of a century, but we have a few years before we practically have to worry about that.

**Review by Jingo88 — DISMISSED** [review]

File: libs/common/src/vault/utils.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+type Year = `${NonZeroIntegers}${NonZeroIntegers}${0 | NonZeroIntegers}${0 | NonZeroIntegers}`;
+
+/**
+ * Takes a string or number value and returns a string value formatted as a valid 4-digit year
+ *
+ * @export
+ * @param {(string | number)} yearInput
+ * @return {*}  {(Year | null)}
+ */
+export function normalizeExpiryYearFormat(yearInput: string | number): Year | null {
```
Comment (djsmith85) [inline libs/common/src/vault/utils.ts:11]: :art: Please add unit tests to check for all the different edge cases.

**Review by djsmith85 — CHANGES_REQUESTED** [review]

File: libs/common/src/vault/utils.ts (hunk 3 above)
Comment (cagonzalezcs) [reply libs/common/src/vault/utils.ts:11]: +1

Comment (jprusik) [reply libs/common/src/vault/utils.ts:11]: Ah, they were updated at the callsite, but I can copy them down here too 👍

Comment (jprusik) [reply libs/common/src/vault/utils.ts:11]: Ah, good call - I actually found another uncovered edge; commit incoming

Comment (jprusik) [reply libs/common/src/vault/utils.ts:11]: 036f16d90c017f27644897650b9a3d66584f1d35

**Review by djsmith85 — APPROVED** [review]

**Review by cagonzalezcs — APPROVED** [review]

**Review by Jingo88 — APPROVED** [review]
